vimiv/imutils/metadata.py

```diff
@@ -0,0 +1,418 @@
+# vim: ft=python fileencoding=utf-8 sw=4 et sts=4
+
+# This file is part of vimiv.
+# Copyright 2017-2021 Christian Karl (karlch) <karlch at protonmail dot com>
+# License: GNU GPL v3, see the "LICENSE" and "AUTHORS" files for details.
+
+"""Utility functions and classes for metadata handling.
+
+All metadata related tasks are implemented in this module. The heavy lifting is done
+using one of the supported metadata libraries, i.e.
+* piexif (https://pypi.org/project/piexif/) and
+* pyexiv2 (https://pypi.org/project/py3exiv2/).
+"""
+
+import contextlib
+import itertools
+from typing import Any, Dict, Tuple, NoReturn, Sequence, Iterable, Optional
+from PyQt5.QtGui import QImageReader
+
+from vimiv.utils import log, lazy, is_hex, files
+
+pyexiv2 = lazy.import_module("pyexiv2", optional=True)
+piexif = lazy.import_module("piexif", optional=True)
+_logger = log.module_logger(__name__)
+
+
+class _InternalKeyHandler(dict):
+    """Handler to load all internal keys of a single image.
+
+    Attributes:
+        _path: Apsolute path of the image to load the metadata of.
+        _reader: QImageReader or None of _path. Accessed via reder property.
+    """
+
+    def __init__(self, path: str):
+        super().__init__(
+            {
+                "vimiv.filesize": ("Vimiv.FileSize", self._get_filesize),
+                "vimiv.xdimension": ("Vimiv.XDimension", self._get_xdimension),
+                "vimiv.ydimension": ("Vimiv.YDimension", self._get_ydimension),
+                "vimiv.filetype": ("Vimiv.FileType", self._get_filetype),
+            }
+        )
+        self._path = path
+        self._reader: Optional[QImageReader] = None
+
+    @property
+    def reader(self) -> QImageReader:
+        """Return QImageReader instance of _path."""
+        if self._reader is None:
+            self._reader = QImageReader(self._path)
+        return self._reader
+
+    def _get_filesize(self):
+        """Get the file size."""
+        return files.get_size_file(self._path)
+
+    def _get_filetype(self):
+        """Get the file type."""
+        return files.imghdr.what(self._path)
+
+    def _get_xdimension(self):
+        """Get the x dimension in pixels."""
+        return self.reader.size().width()
+
+    def _get_ydimension(self):
+        """Get the y dimension in pixels."""
+        return self.reader.size().height()
+
+    def __getitem__(self, key: str) -> Tuple[str, str, str]:
+        """Intrypoint to extract the key of the image at _path.
+
+        Args:
+            key: internal key to fetch.
+        """
+        try:
+            key, func = super().get(key.lower())
+            return (key, key, func())
+        except KeyError:
```

Comment:
If we don't catch the `KeyError` here, the metadata widget can deal with it accordingly and we don't end up with empty entries in the dictionary of the form `"": ("", "")`. This is actually what breaks the tests, it only adds the `no metadata information found` line if the dictionary is completely empty, but false keys are added with empty strings like this.

Reply:
Makes sense. Thank you for spotting that. Though, the metadata widget used `metadata/fetch_keys` which handles `KeyError`. Onl `metadata/fetch_key` throws `KeyError`. So, there is not need to chany anything in metadata widget.

Fixed in 16ec200.